src/Compilers/CSharp/Portable/Binder/Semantics/Operators/BinaryOperatorOverloadResolution.cs

```diff
@@ -825,25 +834,95 @@ private static LiftingResult UserDefinedBinaryOperatorCanBeLifted(TypeSymbol lef
+                    else if (better == BetterResult.Right && candidates[i].Kind != OperatorAnalysisResultKind.Worse)
                     {
                         candidates[i] = candidates[i].Worse();
                     }
                 }
             }
         }
 
-        private BetterResult BetterOperator(BinaryOperatorSignature op1, BinaryOperatorSignature op2, BoundExpression left, BoundExpression right, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        private bool EliminateWorseCandidates_FastPath(
+            BoundExpression left,
+            BoundExpression right,
+            ArrayBuilder<BinaryOperatorAnalysisResult> candidates,
+            ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
-            Debug.Assert(op1.Priority.HasValue == op2.Priority.HasValue);
+            int currentBestIndex = -1;
+            for (int index = 0; index < candidates.Count; index++)
+            {
+                // Find the first applicable candidate
+                if (candidates[index].Kind != OperatorAnalysisResultKind.Applicable)
+                {
+                    continue;
+                }
 
+                // Assume that the current applicable candidate is the best if we don't have any
+                if (currentBestIndex == -1)
+                {
+                    currentBestIndex = index;
+                }
+                else
+                {
+                    var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, left, right, ref useSiteDiagnostics);
+                    if (better == BetterResult.Left)
+                    {
+                        // The current best is better 
+                        candidates[index] = candidates[index].Worse();
+                    }
+                    else if (better == BetterResult.Right)
+                    {
+                        // The current best is worse
+                        candidates[currentBestIndex] = candidates[currentBestIndex].Worse();
+                        currentBestIndex = index;
+                    }
+                    else
+                    {
+                        // The current best is not better
+                        currentBestIndex = -1;
+                    }
+                }
+            }
+
+            // Make sure that every candidate up to the current best is worse
+            for (int index = 0; index < currentBestIndex; index++)
+            {
+                if (candidates[index].Kind == OperatorAnalysisResultKind.Inapplicable)
+                {
+                    continue;
+                }
+
+                var better = BetterOperator(candidates[currentBestIndex].Signature, candidates[index].Signature, left, right, ref useSiteDiagnostics);
+                // Continue if the current best is better 
+                if (better == BetterResult.Left)
+                {
+                    candidates[index] = candidates[index].Worse();
```

Comment:
>candidates[index] = candidates[index].Worse(); [](start = 20, length = 46)

In the EliminateWorseCandidates_SlowPath we are doing this conditionally based on the current kind of the candidate. Should we use the same logic here as well?